Add jasp-stats/jaspBSTS to jasp-desktop as a submodule

vandenman opened the pull request:
Related to https://github.com/jasp-stats/INTERNAL-jasp/issues/1617

Note that the module is not completely finished yet (e.g., there are no unit tests), but @petersen-f is working on that.

vandenman committed:
added jaspBsts

JorisGoosen commented on Desktop/mainwindow.cpp:
```diff
@@ -508,7 +508,7 @@ void MainWindow::loadQML()
 	_ribbonModel->loadModules(	
 		{ 	"jaspDescriptives", "jaspTTests", "jaspAnova", "jaspMixedModels", "jaspRegression", "jaspFrequencies", "jaspFactor" },
-		{ 	"jaspAudit", "jaspBain", "jaspCircular", "jaspDistributions" , "jaspEquivalenceTTests", "jaspJags", "jaspLearnBayes", "jaspMachineLearning",
+		{ 	"jaspAudit", "jaspBain", "jaspCircular", "jaspDistributions" , "jaspEquivalenceTTests", "jaspJags", "jaspLearnBayes", "jaspMachineLearning", "jaspBsts",
```
I thought we were doing the modules in alphabetical order?

JorisGoosen commented:
So this all loos fine except for the order thing, this has to go in for 0.16.1 right?

JorisGoosen reviewed (changes requested):
Sorry it is very nitpicky but if you can change the place where you inserted `jaspBsts` then this can be merged.

vandenman committed:
put list of modules in alphabetical order

JorisGoosen commented:
Still no unit-tests though right?

Well, I see that some new work has been done in https://github.com/petersen-f/jaspBSTS but as the unittests (and perhaps other points) are still missing from jasp-stats/jaspBsts I think it is better to wait until those are in.

Especially given the upcoming feature stop.

JorisGoosen committed:
[WIP] Qt6 - Bringing JASP up to speed (#4677)

* Added the boost:: namespace to all the filesystem. This resolves a conflict when we move to the C++17

* replaced the deprecated QSet::toList()

* Replaced the deprecated `QString::split(QRegEx ...)` with `QString::split(QRegularExpression, ...)`

* Resolved the casting error. I don't know why we need to do get it to QChar, but if we have to, we can do `QChar point = *locale().decimalPoint().toStdString().c_str();`

* This will cause some confusion for the compiler later when the mocing happens. It's a `bool` anyway!

* - Added the missing const operator

* These are not ideal, but C++17 will complain if the types do not match, because `std::max`, `std::min` are template functions. So, we need to consider this. There are more instance of them around which we need to deal with. They are not causing error because they are safer casts but still.

* Replaced the qSort with std::sort. qSort is deprecated, and std::sort is N * Log(N) as well.

* Replaced `qsrand` and `qrand` with `std::srand`, and `std::rand`. Those two are discontinued. If we care, we can do better in c++11/17 regarding the random number

* Using QIODeviceBase instead of `QIODevice` for getting the `OpenModeFlag`. They are not in the `QIODevice` anymore!

* Replaced `midRef` and `leftRef` with `QtString::mid` and `QtString::left`. They are the same and old ones are gone! Btw, they return QString anyway!

* The old one is obsolete

* I'm not so sure about this one, but I think that's you wanted to do. From 6.1, we can use `QLocal::codeToLanguage` to deal with this. We can perhaps replace the entire `isValidLocal` with it since it returns a proper error

* `QDateTime::toTime_t()` is obsolete.

* :exclamation: This is probably not exactly what we want, but getting it right was beyond the focus of this branch. I'm mainly trying to get the 6.2 to work at this point. I'm not even sure what this file does!

* This is an infinite loop. `QStringList.toVector()` spits out `QList<QString>`. Qt Creator picks it up, and shows a warning for it.

I really hope there were a line on top of these file describing their purpose. I have no idea what this file does, or what `fq`, `tq, `fqj` mean!

* Some more minor adaptation. Overall, c++17 is even more strict with types because of the templates and enum classes and Qt 6.2 follows along which is very nice in fact.

- The `UTF-8` is the default option
- The `GL_Lines` is now inside an enum class
- The `fromStdVector` is deprecated. They are relying on `initializer_list` overall which is very nice too. I used a new syntax there, but we will see if we like it or not
- and finally inserted some missing headers

* By the last commit, at least C++ doesn't nag anymore. So, here we are activating the c++17 flag, and moving to the QML side of things. I had to make a few changes to the QMake commands to get a proper c++17 build, but they can be removed or simplified as we move to CMake. I also activated some of the errors to make sure that at this stage I get most of the tricky stuff, of course we can disable them later.

* :warning: Disabled the shadow effects, by removing the `QtGraphicalEffects`. This has moved into the `Qt5Compat.GraphicalEffects` and I expect it to either be implemented somewhere else, or be deprecated soon. We can bring it back if we need to.

* :warning: Replaced the instances of QtQuick.Controls 1.4 with QtQuick.Controls 6.0 wherever was needed! I think Qt is changing their naming scheme, and they are giving their control an umbrella version number which is very nice, but that might cause some issues for us. I think this breaks the SplitView but I cannot immediately see why. I'm marking these commits that we can go back to them, and evaluate our choices

* :exclamation: These are breaking changes, and we need to find a workaround for them

- Styling has changed and this is not valid anymore. I think this might be breaking the AboutWindows but I'm not sure. I feel that's related to the WebEngineQuick. See next commit.
- Commented a bug regarding the shortcut, this should be done differently but I am leaving it for later. I have a list! :P
- I don't think Button has `checkable` and `checked`, maybe it has but Qt doesn't like it. It's not also our Button. I checked!

* :exclamation: Successful build, but broken app!

- Replacing the `QtWebEngine` with `QtWebEngineQuick`. This is explained [here](https://doc.qt.io/qt-6/qtwebengine-changes-qt6.html). One thing that is strange to me is the initialization of the WebEngine. I see Joris' comment, but then [here](https://doc.qt.io/qt-6/qtwebenginequick.html#initialize) it says that it should be initialized before the instantiation of the QApplication.
- I traced back a crash to the `DataSetPackage::index` where that point was causing some termination. I am not 100% sure what's happening there, but I'm not the fan of those pointers. I have tried to deal with it in different ways, but I didn't manage to get it to behave. I think there is something with the enum that breaks it, and leaving it uninitialized at least gives me the build.

* Got it to "compile" and "start"
It doesnt work though

* Fixing some Qml

* some comments on qutils.

Also because QVector is now a subclass/copy of QList certain code wasnt necessary anymore as Amir found out already when he removed:
`inline	std::vector<std::string>			fq (const QStringList						 & vec)		{ return fq(vec.toVector()); }`

* Hmm, this is broken!

but a step in the right direction nonetheless.
The pointer structure of datasetpackage might need to be reconsidered as there are some strange infinite loops going on here.

* Moving to a more sensible structure but CDB is not helping me out by getting stuck reading values from running JASP

I guess I need to boot to linux to get some work done

I guess I should really make it a tree structure

* more attempts but I should just continue tomorrow and rebase+squash this away

* added some submodels to pre-filter datasetpackage before handing it over to the sortfilterproxymodels (which have internal mapping so also filtering rootnodes is impractical)

* maybe using a modified QIdentityProxyModel is better than QAbstractProxyModel as the latter doesnt seem to connect signals and so on.

* alright data is finally being passed through and JASP doesnt seem to crash.
Except if I turn on the nice custom rendering lines in DataSetView...
So I've disabled them for now.

* label can now be selected

* Alright, the values and labels are gotten properly now and some more qml warnings silenced

* Enabling and fixing checked and checkable in VariablesWindow

* almost got the splitviews to behave again as well

* something weird is going on with the splithandlers

* Results and data panel are now accessible and visible

* More functionparameters for onSignal handlers

* make sure IPCChannel is always created *before* starting JASPEngine

* Implements EJs request for moving forms a bit out of screen lefthand if wanted

Make EJs desire work better

* - Remove old unused file BasicTableViewJASP
- make sure suggested column icons in variableslists actually change with the theme.
- undo some breaking changes by Amir

* Put shadows back

* Apparently the lines in DataSetView work again so I can enable them then

* Convert shaders to separate files so they can later be converted into .qsb
This can be easily done through cmake as described here: https://doc.qt.io/qt-6/qtshadertools-build.html

Rewrite shaders to work with qsb

for now, until cmake is done, I compiled them as:

`D:\Broncode\JASP\jasp-desktop\Desktop\components\JASP\Widgets>D:\Qt\6.2.1\msvc2019_64\bin\qsb  --glsl 100es,120,150 --hlsl 50 --msl 12 -b -o ImageInverter.vert.qsb ImageInverter.vert`
`D:\Broncode\JASP\jasp-desktop\Desktop\components\JASP\Widgets>D:\Qt\6.2.1\msvc2019_64\bin\qsb  --glsl 100es,120,150 --hlsl 50 --msl 12 -o ImageInverter.frag.qsb ImageInverter.frag`

* Fixing some qml (arguments for handlers)

* Make special plot-scheme-handler to avoid annoying "not allowed to load local resources" problem with Qt6

* Fixes and refactoring

- no need for singleShotTime with queued connection
- also some changes in Qt6 broke `source: "repeatedMeasures"` as opposed to `source: ["repeatedMeasures"]`, ive added a workaround.

* Make sure data is hidden if a jasp file including analyses is loaded

* AnalysisFormExpander now opens smoothly again

* new major R-Interface number

* Up minor version, it is now at 0.16.2

Co-authored-by: Amir Masoud Abdol <amirmasoudabdol@icloud.com>

JorisGoosen committed:
Use fix for https://github.com/jasp-stats/INTERNAL-jasp/issues/1735 in jaspResults (more flexible mustContain)
small qml fix

JorisGoosen committed:
Make sure it uses Qt 6.2.3 by default

and macos on mac

boutinb committed:
Remove TypeError in CheckBox and RadioButton

With Qt6, the CheckBoxes and RadioButtons generated TypeErrors due to apparently https://forum.qt.io/topic/131823/lots-of-typeerrors-in-console-when-migrating-to-qt6

weblate committed:
Translations update from Hosted Weblate (#4718)

* Translated using Weblate (Japanese)

Currently translated at 100.0% (768 of 768 strings)

Translation: JASP/jaspDesktop
Translate-URL: https://hosted.weblate.org/projects/jasp/jaspdesktop/ja/

* Translated using Weblate (French)

Currently translated at 100.0% (768 of 768 strings)

Translation: JASP/jaspDesktop
Translate-URL: https://hosted.weblate.org/projects/jasp/jaspdesktop/fr/

* Translated using Weblate (Indonesian)

Currently translated at 25.9% (199 of 768 strings)

Translation: JASP/jaspDesktop
Translate-URL: https://hosted.weblate.org/projects/jasp/jaspdesktop/id/

Co-authored-by: Daiki Hojo <dhojo.bayes@gmail.com>
Co-authored-by: Jordy STEFAN <jordy.stefan@askoria.eu>
Co-authored-by: Reza Almanda <rezaalmanda27@gmail.com>

Commit:
updated translation files

vandenman commented:
@JorisGoosen happy to merge whenever you are! The unit tests are now fixed in the module.

JorisGoosen commented:
Aaah, I wanted to merge this but I think it is better to wait until the meeting tomorrow where we discuss 0.16.2

vandenman commented:
Didn't even know we were planning 0.16.2 but sounds like a good plan!

EJWagenmakers commented:
Definitely a good plan!

vandenman commented:
closed in favor of https://github.com/jasp-stats/jasp-desktop/pull/4764
